packages/contractkit/src/identity/offchain/schema-utils.ts

```diff
@@ -21,62 +29,399 @@ export class OffchainError extends RootError<SchemaErrorTypes.OffchainError> {
   }
 
-  try {
-    const asJson = JSON.parse(resp.result)
-    const parseResult = type.decode(asJson)
-    if (isRight(parseResult)) {
-      return Ok(parseResult.right)
+  async readAsResult(account: string): Promise<Result<Buffer, SchemaErrors>> {
+    const rawData = await this.wrapper.readDataFromAsResult(
+      account,
+      (buf) => buildBinaryEIP712TypedData(this.wrapper, this.dataPath, buf),
+      this.dataPath
+    )
+    if (!rawData.ok) {
+      return this.readEncrypted(account)
     }
-    return Err(new RootError(SchemaErrorTypes.InvalidDataError))
-  } catch (error) {
-    return Err(new RootError(SchemaErrorTypes.InvalidDataError))
+
+    const encryptedResult = await this.readEncrypted(account)
+
+    if (encryptedResult.ok) {
+      return encryptedResult
+    }
+    return Ok(rawData.result)
+  }
+
+  read = makeAsyncThrowable(this.readAsResult.bind(this))
+
+  private async readEncrypted(account: string) {
+    return readEncrypted(this.wrapper, this.dataPath, account)
   }
 }
 
-export const readWithSchema = makeAsyncThrowable(readWithSchemaAsResult)
+// label = PRF(ECDH(A, B), A || B || data path)
+// ciphertext path = "/cosmetic path/" || base64(label)
+export function getDataPath(
+  path: string,
+  sharedSecret: Buffer,
+  senderPublicKey: string,
+  receiverPublicKey: string
+) {
+  const senderPublicKeyBuffer = Buffer.from(trimLeading0x(senderPublicKey), 'hex')
+  const receiverPublicKeyBuffer = Buffer.from(trimLeading0x(receiverPublicKey), 'hex')
+
+  return createHmac('sha3-256', sharedSecret)
+    .update(Buffer.concat([senderPublicKeyBuffer, receiverPublicKeyBuffer, Buffer.from(path)]))
+    .digest('base64')
+}
 
-export const writeWithSchema = async <T>(
+// Assumes that the wallet has the dataEncryptionKey of wrapper.self available
+// TODO: Should check and throw a more meaningful error if not
+export const writeEncrypted = async (
   wrapper: OffchainDataWrapper,
-  type: t.Type<T>,
   dataPath: string,
-  data: T
+  data: Buffer,
+  toAddress: string
 ) => {
-  if (!type.is(data)) {
-    return
+  const accounts = await wrapper.kit.contracts.getAccounts()
+  const [fromPubKey, toPubKey] = await Promise.all([
+    accounts.getDataEncryptionKey(wrapper.self),
+    accounts.getDataEncryptionKey(toAddress),
+  ])
+  const wallet = wrapper.kit.getWallet()
+  const sharedSecret = await wallet.computeSharedSecret(publicKeyToAddress(fromPubKey), toPubKey)
+
+  const computedDataPath = getDataPath(dataPath, sharedSecret, fromPubKey, toPubKey)
+  const encryptedData = Encrypt(
+    Buffer.from(trimPublicKeyPrefix(toPubKey), 'hex'),
+    Buffer.from(data)
+  )
+
+  const prefixedDataPath = '/ciphertexts/' + computedDataPath
+  const signature = await signBuffer(wrapper, prefixedDataPath, encryptedData)
+  await wrapper.writeDataTo(encryptedData, signature, prefixedDataPath)
+}
+
+//  file_ciphertext = IV || E(message key, IV, data)
+export const writeEncryptedWithSymmetric = async (
```

Comment:
We probably not export this method given it's interface complexity.
```suggestion
const writeEncryptedWithSymmetric = async (
```